tests/MongoDB.Analyzer.Tests/Infrastructure/ClassBasedTestCaseSourceAttribute.cs

```diff
@@ -0,0 +1,90 @@
+﻿// Copyright 2021-present MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License")
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using MongoDB.Analyzer.Core;
+using MongoDB.Analyzer.Tests.Common;
+using NuGet.Versioning;
+
+namespace MongoDB.Analyzer.Tests.Infrastructure;
+
+public sealed class ClassBasedTestCasesSourceAttribute : Attribute, ITestDataSource
+{
+    private readonly Type _testCasesProviderType;
+
+    public ClassBasedTestCasesSourceAttribute(Type testCasesProviderType)
+    {
+        _testCasesProviderType = testCasesProviderType;
+    }
+
+    public IEnumerable<object[]> GetData(MethodInfo methodInfo)
+    {
+        var testCasesMethods = _testCasesProviderType
+            .GetMembers(BindingFlags.Public | BindingFlags.Instance | BindingFlags.Static)
+            .Where(m => m.MemberType == MemberTypes.Method &&
+                m.DeclaringType == _testCasesProviderType &&
+                m.GetCustomAttributes().Any(a => a is JsonDiagnosticTestCaseAttribute));
+
+        var data = testCasesMethods.SelectMany(m =>
+            CreateTestCases(m).Select(t => new object[] { t }).ToArray()).ToArray();
+
+        return data;
+    }
+
+    public string GetHighestDeclaringType(MemberInfo memberInfo)
+    {
+        Type declaringType = memberInfo.DeclaringType;
+
+        while (declaringType.DeclaringType != null)
+        {
+            declaringType = declaringType.DeclaringType;
+        }
+
+        return declaringType.FullName;
+    }
+
+    public string GetDisplayName(MethodInfo methodInfo, object[] data)
+    {
+        var testCase = (DiagnosticTestCase)data[0];
+        var linqVersion = testCase.LinqVersion == Common.LinqVersion.V3 ? "V3" : testCase.LinqVersion == Common.LinqVersion.Undefined ? "U" : "";
+        return $"v{testCase.Version}_{linqVersion}_{testCase.MethodName}";
+    }
+
+    private DiagnosticTestCase[] CreateTestCases(MemberInfo memberInfo)
+    {
+        var testCasesAttributes = memberInfo
+            .GetCustomAttributes()
+            .OfType<JsonDiagnosticTestCaseAttribute>();
+
+        var fileName = PathUtilities.GetTestCaseFileFullPathFromName(GetHighestDeclaringType(memberInfo));
```

Comment:
Can this be deduced from `_testCasesProviderType` once?

Reply:
Yes made the changes